Invalid C code generation with `iterator`/`nimvm`

### Nim Version

```
Nim Compiler Version 2.2.8 [Linux: amd64]
Compiled at 2026-04-09
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: 4f500679b196fad944caa50a753f5bbfaefda001
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.9 [Linux: amd64]
Compiled at 2026-04-09
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: c7339047164d8c1feeef6822ba3c8ab4085db522
active boot switches: -d:release
```

```
Nim Compiler Version 2.3.1 [Linux: amd64]
Compiled at 2026-04-09
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: 188aa1714e0bd95caca408265d49c720ae739449
active boot switches: -d:release
```

### Description

```nim
iterator c(): int =
  when nimvm: yield 0
  else: yield 1
for w in c():
  let n = w
  (proc() = discard n)()
```

### Current Output

```text
......................................................................
CC: system/exceptions.nim
CC: std/private/digitsutils.nim
CC: system/dollars.nim
CC: system.nim
CC: j.nim
/tmp/tmp.R39w5VGhoB/@mj.nim.c: In function ‘colonanonymous___j_u4’:
/tmp/tmp.R39w5VGhoB/@mj.nim.c:72:7: error: ‘n__j_u3’ undeclared (first use in this function); did you mean ‘n__j_u6’?
   72 | (void)n__j_u3;
      |       ^~~~~~~
      |       n__j_u6
/tmp/tmp.R39w5VGhoB/@mj.nim.c:72:7: note: each undeclared identifier is reported only once for each function it appears in
```

### Expected Output

```text
No invalid C code generation
```

### Known Workarounds

_No response_

### Additional Information

_No response_

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -336,7 +336,7 @@ proc introduceNewLocalVars(c: PTransf, n: PNode): PNode =
     if a.kind == nkSym:
       n[1] = transformSymAux(c, a)
     return n
-  of nkProcDef, nkFuncDef, nkMethodDef, nkConverterDef: # todo optimize nosideeffects?
+  of nkLambdaKinds, nkProcDef, nkFuncDef, nkMethodDef, nkConverterDef: # todo optimize nosideeffects?
     result = newTransNode(n)
     let x = newSymNode(copySym(n[namePos].sym, c.idgen))
     c.transCon.mapping[n[namePos].sym.itemId] = x

--- a/tests/iter/titer_issues.nim
+++ b/tests/iter/titer_issues.nim
@@ -457,3 +457,12 @@ let runes1 = buggyVersion("en")  # <-- CRASHES HERE
 
 doAssert runes1.len == runes2.len
 # echo "Got ", runes1.len, " runes"
+
+
+block: # bug #25724
+  iterator c(): int =
+    when nimvm: yield 0
+    else: yield 1
+  for w in c():
+    let n = w
+    (proc() = discard n)()

